File: ruby/07_10001st_prime.rb
# By listing the first six prime numbers: 2, 3, 5, 7, 11, and 13, we can see that the 6th prime is 13.
#
# What is the 10 001st prime number?

def nthPrime(n)

    return 2 if n == 1

    count, number = 1, 3

    loop do

        if isPrime(number)
            count += 1
        end

        return number if count == n

        number += 2
    end
end

def isPrime(number)
    return false if number < 2
    return true if number < 4
    return false if number % 2 == 0
    return true if number < 9
    return false if number % 3 == 0

    factor = 5
    until factor * factor > number
        return false if number % factor == 0 || number % (factor + 2) == 0
        factor += 6
    end

    true
end

p(nthPrime(10001))

File: javascript/07_10001st_prime.js
// By listing the first six prime numbers: 2, 3, 5, 7, 11, and 13, we can see that the 6th prime is 13.
//
// What is the 10 001st prime number?

function nthPrime(n) {

    if (n === 1) return 2;

    var i = 3,
        count = 1;

    while (true) {

        if (isPrime(i)) count++;

        if (count === n) return i;

        i += 2;
    }
}

function isPrime(n) {
    if (n < 2) return false;
    if (n < 4) return true;
    if (n % 2 === 0) return false;
    if (n < 9) return true;
    if (n % 3 === 0) return false;

    var i = 5, limit = Math.floor(Math.sqrt(n));

    while (i <= limit) {
        if (n % i === 0 || n % (i + 2) === 0) return false;
        i += 6;
    }

    return true;
}

console.log(nthPrime(10001));
